src/Model/Qualification.php

```diff
@@ -44,6 +44,32 @@ public function undefined( string $part ): bool {
 		return self::UNDEFINED === $this->$part;
 	}
 
+	public function isApproximate(): bool {
+		return in_array( $this->year, [ self::APPROXIMATE, self::UNCERTAIN_AND_APPROXIMATE ] )
+			|| in_array( $this->month, [ self::APPROXIMATE, self::UNCERTAIN_AND_APPROXIMATE ] )
+			|| in_array( $this->day, [ self::APPROXIMATE, self::UNCERTAIN_AND_APPROXIMATE ] );
+	}
+
+	public function isUncertain(): bool {
+		$approximate = [ self::UNCERTAIN, self::APPROXIMATE, self::UNCERTAIN_AND_APPROXIMATE ];
+		return in_array( $this->year,  $approximate)
+			|| in_array( $this->month, $approximate )
+			|| in_array( $this->day, $approximate );
+	}
+
+	/**
+	 * @return array<string, array<string> >
+	 */
+	public function getUncertainty(): array {
```

Comment:
I'm not sure what `array<string>` actually means. Is string about the key or the value? Ideally use a more common form, like `string[]` or `array<key, value>`

Style remark: there should be no space between the closing `>`

Reply:
> Method EDTF\Model\Qualification::getUncertainty() should return array<string, array<string>> but returns array<int, array<int, string>>

This error is correct. Right now the keys of the top level array are of type int. They are the year, month and day values, coming from the fields, which are all int. So the doc is wrong.

Btw the code in `getUncertainty` and in its user (`composeMessage`) looks overly complicated. I strongly suspect it can be simpler, and without passing around this array data structure. Passing around this type of array data structure also happens to be an anti-pattern. Can't the code work with `Qualification` itself? That already is the data structure, with the benefit of having clearly defined fields, better typing options, and the ability to glue methods on top of the data.

So my recommendation is to try ditch the newly added `getUncertainty`.

Reply:
Don't be afraid to have a line of code that deals with year, another with month and a third with day. No need to put everything in a loop for the sake of DRY. See https://www.entropywins.wtf/blog/2017/09/06/the-fallacy-of-dry/

Reply:
> This error is correct. Right now the keys of the top level array are of type int. They are the year, month and day values, coming from the fields, which are all int. So the doc is wrong.

Reply:
> This error is correct. Right now the keys of the top level array are of type int. They are the year, month and day values, coming from the fields, which are all int. So the doc is wrong.

right, I will check with <int, array<int, string>>, that was my mistake.
So do you think composeMessage can directly use the values from the class Qualification ? I can try, however now getUncertainty distributes each part of the date to an array for each uncertainty kind, and, with reference to the loop,  for certain cases the nested array is not provided for a simpler use of composeMessage.
I will check if I can move this data-structure from "method-level" to "class-level" which I think can be grasped from your recommendation